### PR TITLE
WIP - straw man of exposing hierarchical file input snapshots to operation listeners

### PR DESCRIPTION
**Pull request by ldaley**
Put up to identify issues before finishing the real implementation.

Not to be merged.

### CODE REVIEW

File: subprojects/core/src/main/java/org/gradle/api/internal/changedetection/state/mirror/logical/DefaultFileCollectionFingerprint.java [hunk 1]
```diff
@@ -38,20 +38,22 @@
...
     private HashCode hash;
 
     public DefaultFileCollectionFingerprint(FingerprintCompareStrategy strategy, Map<String, NormalizedFileSnapshot> snapshots, @Nullable HashCode hash) {
-        this(strategy, snapshots, hash, null);
+        this(null, strategy, snapshots, hash, null);
     }
 
     public DefaultFileCollectionFingerprint(FingerprintingStrategy strategy, Iterable<PhysicalSnapshot> roots) {
-        this(strategy.getCompareStrategy(), strategy.collectSnapshots(roots), null, roots);
+        this(strategy.getIdentifier(), strategy.getCompareStrategy(), strategy.collectSnapshots(roots), null, roots);
```
Comment (wolfs) [inline subprojects/core/src/main/java/org/gradle/api/internal/changedetection/state/mirror/logical/DefaultFileCollectionFingerprint.java]: `CompareStrategy` is now an enum. Would that be enough for the scan plugin? Then it wouldn't be possible to distinguish between compile and runtime classpath.

File: subprojects/core/src/main/java/org/gradle/api/internal/tasks/execution/ResolveBuildCacheKeyExecuter.java [hunk 2]
```diff
@@ -144,6 +150,145 @@ public String apply(HashCode input) {
...
+                for (PhysicalSnapshot tree : fileCollectionSnapshot.getRoots()) {
+                    if (tree.getType() == FileType.Missing) {
+                        continue;
+                    }
+
+                    state.path = tree.getAbsolutePath();
+                    state.name = tree.getName();
+                    state.hash = null;
+
+                    visitor.preRoot(state);
```
Comment (wolfs) [inline subprojects/core/src/main/java/org/gradle/api/internal/tasks/execution/ResolveBuildCacheKeyExecuter.java]: On `master`, there is now only an `accept` method on `FileCollectionSnapshot`. That method would visit all roots in order. The visitor here would need to make sure it detects when we switch roots. Do you think that is feasible?

For checking if something is a root the visitor would need to keep track of the depth it currently is by tracking `preVisitDirectory`/`postVisitDirectory` calls.

File: subprojects/core/src/main/java/org/gradle/api/internal/changedetection/state/DefaultCompileClasspathSnapshotter.java [hunk 3]
```diff
@@ -40,7 +41,7 @@ public DefaultCompileClasspathSnapshotter(ResourceSnapshotterCacheService cacheS
     public FileCollectionSnapshot snapshot(FileCollection files, PathNormalizationStrategy pathNormalizationStrategy, InputNormalizationStrategy inputNormalizationStrategy) {
         return super.snapshot(
             files,
-            new ClasspathFingerprintingStrategy(IGNORE, classpathResourceHasher, cacheService));
+            new ClasspathFingerprintingStrategy(FingerprintingStrategy.Identifier.COMPILE_CLASSPATH, IGNORE, classpathResourceHasher, cacheService));
```
[the remark this answers is not in the document]
Comment (ldaley) [reply subprojects/core/src/main/java/org/gradle/api/internal/changedetection/state/DefaultCompileClasspathSnapshotter.java]: Yes, they do. The trees are different due to the different logical values for the non class files.

File: subprojects/core/src/main/java/org/gradle/api/internal/changedetection/state/mirror/logical/DefaultFileCollectionFingerprint.java (hunk 1 above)
Comment (ldaley) [reply subprojects/core/src/main/java/org/gradle/api/internal/changedetection/state/mirror/logical/DefaultFileCollectionFingerprint.java]: No, we need the extra detail.

**Comment by ldaley** [comment]
@wolfs made that change to just use a visitor. Seems to work ok.

https://github.com/gradle/gradle/pull/5944/files#diff-cdd6f017e5b1e29aca749b5ee8c9efe7R221

**Pull request closed without merging**